kelpie-stream10-jegor

JegorKuz opened the pull request.

JegorKuz committed:
[UPDATE] url-shortener
1)Настроил HashCache(Redis)
2)Настроил UrlHashCache(Redis)
3)Настроил PreparedUrlHashRepository(PostgreSQL)
4)Настроил UrlRepository(Cassandra)
5)Создал HashGenerator
6)Создал Base62Encoder
7)Создал HashGeneratorScheduler
8)Вроде настроил Redis Pub/Sub
9)Добавил Controller/Service
10)Настроил docker-compose + application.yaml
11)Настроил конфиг для многопоточности

JegorKuz committed:
[UPDATE] url-shortener
Вся логика работает от получения короткой ссылки до получения из короткой полной.

JegorKuz committed:
[UPDATE] url-shortener
1) Fixed HashCache async generation.
2) Full logic of retrieving FullUrl from Hash moved into UrlHashCache.

JegorKuz committed:
[UPDATE] url-shortener
1) Fixed HashCache async generation.
2) Full logic of retrieving FullUrl from Hash moved into UrlHashCache.
3) Исправил подключение обоих кэшей к одному порту из-за чего UrlHashCache работал не верно.

JegorKuz committed:
[UPDATE] url-shortener
Исправил Listener для сообщений об удалении сущностей из UrlHashCache. Теперь после удаления из кэша сущность в Cassandra меняет full_url на nul, а так же сущность в PostgreSQL меняет taken на false.

miked4king commented on src/main/java/faang/school/urlshortenerservice/ServiceTemplateApplication.java:
```diff
@@ -1,18 +1,41 @@
+@EnableAsync
 @EnableFeignClients("faang.school.urlshortenerservice.client")
+@EnableCassandraAuditing
+@ConfigurationPropertiesScan
```
ServiceTemplateApplication ?

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,8 @@
+package faang.school.urlshortenerservice.controller;
+
+public interface UrlController {
```
убрать

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlControllerImpl.java:
```diff
@@ -0,0 +1,30 @@
+    private final UrlService urlService;
+
+    @GetMapping
+    public String getFullUrl(@RequestParam String shortUrl) {
```
dto для запроса/ответа

JegorKuz replied:
Для запроса да, для ответа мне кажется излишне?
Один делает RedirectView, второй просто возвращает хэш для этого ведь и стринги хватит?

miked4king replied:
да, для ответа не нужно

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlControllerImpl.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/url")
```
/api/v1 - в конфиг
/url - для post метода

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/HashCache.java:
```diff
@@ -0,0 +1,58 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class HashCache {
```
интерфейс ?

JegorKuz replied:
С интерфейсом напряжно, у них всё таки разные get/put
HashCache:
String get()
Long put(Set<String> hashes)
UrlHashCache:
String get(String hash)
void put(String hash, String fullUrl)

miked4king replied:
сделать два интерфейса мб?

miked4king commented on src/main/java/faang/school/urlshortenerservice/repository/postgre/PreparedUrlHashRepository.java:
```diff
@@ -0,0 +1,39 @@
+                WHERE p.taken = false
+                LIMIT :limit
+            """, nativeQuery = true)
+    Set<String> findUntakenHashes(int limit);
```
название метода подумаем

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+        if (hashCache.getCapacity() / 10 > hashCache.getCurrentSize()) {
+            log.info("{}/{}", hashCache.getCapacity(), hashCache.getCurrentSize());
+            hashGeneratorScheduler.triggerGeneration();
+        }
```
вынесем в hashCache
hashCache.getCapacity() / 10 > hashCache.getCurrentSize() - давай в метод с информативным названием

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+    private String shortenFullUrl(String hash) {
+        return "mylink.com/" + hash;
+    }
```
mylink.com/ - либо в кфг, либо в контроллере принять HttpServletRequest и из него вытащить

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+        String shortUrl = shortenFullUrl(hash);
+
+        // Cassandra UPSERT UPdate or inSERT
```
комменты в итоговом виде убрать

miked4king commented on src/main/java/faang/school/urlshortenerservice/entity/UrlHash.java:
```diff
@@ -0,0 +1,33 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
уберем

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+                .hash(hash)
+                .build());
+        // UrlHashCache
+        urlHashCache.put(hash, fullUrl, 120);
```
120 ?

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/UrlHashCache.java:
```diff
@@ -0,0 +1,70 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class UrlHashCache {
```
интерфейс

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/UrlHashCache.java:
```diff
@@ -0,0 +1,70 @@
+    private static final String URL_MAPPINGS_COUNT_KEY = "url_mappings_count";
+
+    private final UrlHashRepository urlHashRepository;
+    @Qualifier("urlHashCacheRedisTemplate")
```
@Qualifier("urlHashCacheRedisTemplate") - обычно в конструкторе, можешь почитать что-нибудь на эту тему

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/UrlHashCache.java:
```diff
@@ -0,0 +1,70 @@
+    @Qualifier("urlHashCacheRedisTemplate")
+    private final RedisTemplate<String, String> urlHashCacheRedisTemplate;
+
+    public long getSize() {
```
не используется

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/UrlHashCache.java:
```diff
@@ -0,0 +1,70 @@
+        ValueOperations<String, String> valueOps = urlHashCacheRedisTemplate.opsForValue();
+
+        Boolean isNewKey = valueOps.setIfAbsent(hash, fullUrl, Duration.ofSeconds(ttlSeconds)); // сюда брать из properties.ttl? или настройка в application.yaml?
+        log.info("DEBUG: Using RedisTemplate with ConnectionFactory type: {}", urlHashCacheRedisTemplate.getConnectionFactory().getClass().getName());
```
общий коммент убрать лишнее)

miked4king commented on src/main/java/faang/school/urlshortenerservice/cache/UrlHashCache.java:
```diff
@@ -0,0 +1,70 @@
+        } else {
+            log.info("Hash: {} not found in URL mappings cache (might have expired).", hash);
+            // Если не в кэше, ищем в Cassandra
+            fullUrl = urlHashRepository.findByHash(hash);
```
findByHash вернет Optional ?

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlControllerImpl.java:
```diff
@@ -0,0 +1,30 @@
+    private final UrlService urlService;
+
+    @GetMapping
+    public String getFullUrl(@RequestParam String shortUrl) {
```
реализовать редирект

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        return null;
+    }
```
убираем

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+@RequiredArgsConstructor
+public class UrlServiceImpl implements UrlService {
+
+    private static final String DOMAIN_URL = "mylink.com/";
```
убрать

miked4king commented on src/main/java/faang/school/urlshortenerservice/scheduler/HashGenerator.java:
```diff
@@ -0,0 +1,105 @@
+package faang.school.urlshortenerservice.scheduler;
```
почему в scheduler лежит HashGenerator ?

miked4king commented on src/main/java/faang/school/urlshortenerservice/repository/postgre/PreparedUrlHashRepository.java:
```diff
@@ -0,0 +1,39 @@
+package faang.school.urlshortenerservice.repository.postgre;
```
postgres

miked4king commented on src/main/java/faang/school/urlshortenerservice/scheduler/HashGenerator.java:
```diff
@@ -0,0 +1,105 @@
+        } else {
+            log.info("No hashes were added to the cache");
+            // Do a rollback?
+        }
```
?

miked4king commented on src/main/java/faang/school/urlshortenerservice/scheduler/HashGenerator.java:
```diff
@@ -0,0 +1,105 @@
+                        .build())
+                .collect(Collectors.toSet());
+
+        try {
```
можно без try/catch

miked4king commented on src/main/java/faang/school/urlshortenerservice/scheduler/HashGeneratorScheduler.java:
```diff
@@ -0,0 +1,52 @@
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class HashGeneratorScheduler {
```
как будто можем отказаться

miked4king commented on src/main/java/faang/school/urlshortenerservice/scheduler/HashGeneratorScheduler.java:
```diff
@@ -0,0 +1,52 @@
+        triggerCacheFilling();
+    }
+
+    @Scheduled(cron = "${cron.hash_generator.every_hour}")
```
https://www.baeldung.com/shedlock-spring - почитай для корп разработки

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/hash_cache/HashCacheRedisConfiguration.java:
```diff
@@ -0,0 +1,42 @@
+package faang.school.urlshortenerservice.config.redis.hash_cache;
```
название пакета 1 слово - существительное

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/hash_cache/RedisHashCacheProperties.java:
```diff
@@ -0,0 +1,17 @@
+package faang.school.urlshortenerservice.config.redis.hash_cache;
```
либо оставить в пакете redis

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/hash_cache/RedisHashCacheProperties.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Component
```
?

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/hash_cache/RedisHashCacheProperties.java:
```diff
@@ -0,0 +1,17 @@
+    private long batchSize;
+    private int capacity;
+    int port;
+    String host;
```
модификаторы

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/url_hash_cache/RedisUrlHashCacheProperties.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+@ConfigurationProperties(prefix = "spring.data.redis.url-hash-cache")
+public class RedisUrlHashCacheProperties {
+    String key;
```
модификаторы

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/url_hash_cache/RedisUrlHashCacheProperties.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Component
```
?

miked4king commented on src/main/java/faang/school/urlshortenerservice/utils/Base62Encoder.java:
```diff
@@ -0,0 +1,53 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
@UtilityClass

miked4king commented on src/main/java/faang/school/urlshortenerservice/utils/Base62Encoder.java:
```diff
@@ -0,0 +1,53 @@
+     * @param requiredLength Требуемая длина выходной строки.
+     * @return Строка Base62 заданной длины.
+     */
+    public String encode(long num, int requiredLength) {
```
на будущее ищи сначала готовые либы

JegorKuz replied:
поискал, нашел несколько, но не нашел либу которая берёт как аргумент длинну генерируемого хеша, суть рекомендации понял, на будущее учту, но пока и такая сойдёт ведь?

miked4king replied:
да, можем оставить, оставил на будущее)

miked4king commented on src/main/java/faang/school/urlshortenerservice/utils/Base62Encoder.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    public String encode(long num, int requiredLength) {
+        if (num < 0) {
+            throw new IllegalArgumentException("Number must be non-negative.");
```
старайся кастомные исключения бросать

miked4king reviewed (changes requested):
обработчик ошибок хотелось бы увидеть

JegorKuz committed:
[UPDATE] url-shortener
1) Почистил логи от дебага.
2) TTL настроен UrlHashCache вместо ручного присваивания в сервисе.
3) Заменил RequestParam на DTO.

JegorKuz commented:
Не получиолсь вынести запрос на генерацию доп хэшей в кэш с ними, всё равно возникает цикличная зависимость, получилось разве что убрать Scheduler.
HashCache -> HashCacheFiller(бывший HashGeneratorScheduler) -> HashGenerator -> HashCache
самое простое оставить одну строчку в сервисе.

JegorKuz committed:
[UPDATE] url-shortener
1) Добавил обработчик ошибок.
2) Заменил возвращаемый тип на RedirectView в контроллере.
3) Добавил @Valid для проверки полей RequestDto.
4) Поправил помарки.

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/redis/url_hash_cache/UrlHashCacheRedisConfiguration.java:
```diff
@@ -0,0 +1,74 @@
+        container.setConnectionFactory(connectionFactory);
+        // Подписываемся на канал событий истечения срока действия ключей
+        // __keyevent@<db>__:expired
+        // @0 означает базу данных по умолчанию (DB 0).
```
комменты убрать или оформить как javadoc

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlControllerImpl.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @PostMapping("${spring.controller.api}${spring.controller.version}/url")
+    public String createShortUrl(@Valid @RequestBody ShortUrlRequestDto dto) {
```
тут для ответа сделать dto)

miked4king replied:
мне кажется название класса ShortUrlRequestDto должно быть в методе redirectToOriginalUrl, т.к мы принимаем короткий url, а FullUrlRequestDto тут

miked4king commented on src/main/java/faang/school/urlshortenerservice/handler/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,83 @@
+        );
+    }
+
+    // Как будто пользователю похер? Он ведь даже этого никогда не увидит?
```
кстати да, у тебя же генерация ассинхронно выполняется, смысла нет в выбросе исключения, достаточно будет залоггировать (обработчик ниже тоже)
если по хорошему сделать метрику, тоесть можно настроить так, чтобы у тебя инкрементилась метрика и приходили уведомления в мессенджер например, чтобы оперативно понимать когда происходят сбои в сервисе в критических местах.

JegorKuz replied:
Добавить добавил, но там надо дальше настраивать что-бы отсылать в какой-нибудь графана, это уже история на будущее, сейчас news feed.

JegorKuz committed:
[UPDATE] url-shortener
Исправил помарки
